zhmcclient/_cpc.py

```diff
@@ -2200,6 +2200,262 @@ def single_step_install(
+        """
+        Installs and activates firmware on the Support Element (SE) of this CPC.
+
+        Brief summary of firmware levels:
+
+            The firmware is segmented into different subsystems called
+            "Engineering Change" (EC), or sometimes "EC stream". An EC stream
+            is identified by an EC number (e.g. "P30719").
+
+            Each EC stream is at a particular code level called "Microcode
+            Level" (MCL), or sometimes "MCL level". An MCL level for an EC
+            stream is identified by an MCL number (e.g. "001"). MCL numbers
+            are unique only within their EC stream and are consecutive numbers
+            that increase towards newer MCL levels. Updates within an EC stream
+            are installed sequentially, so if an EC stream has a particular
+            MCL level installed, that implies that all earlier MCL levels
+            within that EC stream are also installed.
+
+            A particular set of MCL numbers for each EC stream is collected
+            into a "bundle level" for the SE (e.g. "S81") or HMC (e.g. "H20").
+
+        The firmware level for this method can be specified in three ways:
+
+        * By specifying `bundle_level`: The updates for the specified bundle
+          level will be installed.
+        * By specifying `ec_levels`: Specific MCL levels for the specified EC
+          streams will be installed.
+        * By not specifying `bundle_level and `ec_levels`: All locally
+          available updates will be installed.
+
+        In all cases, the updates to be installed must already be available on
+        the SE; they are *not* automatically downloaded from the IBM support
+        site or from an FTP server.
+
+        This method is implemented by performing the "CPC Install and Activate"
+        operation which performs the following steps:
+
+        * The specified updates are installed.
+        * If all updates are installed successfully, they are activated, which
+          includes rebooting the SE of this CPC.
+
+        If an error occurs when installing the updates, any updates that were
+        successfully installed are rolled back.
+
+        Note that this operation does *not* perform a backup, an accept of
+        previously activated updates, or an accept of the newly installed
+        updates.
+
+        Note that this operation does not require that previously activated
+        updates are first accepted before invoking this operation.
+
+        Note that it is not possible to downgrade the SE firmware with this
+        operation.
+
+        Authorization requirements:
+
+        * Object-access permission to this CPC.
+        * Task permission to the "Change Internal Code" task.
+
+        Parameters:
+
+          bundle_level (string): Name of the bundle to be installed on the SE
+            of this CPC (e.g. 'S51').
+
+            This parameter is mutually exclusive with `ec_levels`. If both
```

Comment:
The code below does not check the mutual exclusive requirement. Is it the design point of zhmcclient to let this pass and let the underlying WSAPI implementation detect this?

Reply:
Yes, that is done for most operations in the zhmcclient. The idea is not to double check because this would shorten only the error path, but make the good path slightly (very minimally though) longer.